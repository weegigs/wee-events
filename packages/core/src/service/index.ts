export * from "./state";
export * from "./dispatcher";
export * from "./loader";
export * from "./service";
