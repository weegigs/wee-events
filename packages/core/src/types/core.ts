import * as z from "zod";

export type Primitive = string | number | boolean;
export type Value = Primitive | Payload | List;
export type EmptyPayload = { [member: string]: never };
export type Payload = { [member: string]: Value } | EmptyPayload;
export type List = Primitive[] | Payload[] | List[];

export namespace Payload {
  const primitive = z.union([z.string(), z.number(), z.boolean()]);
  const list: z.Schema<List> = z.lazy(() => z.union([z.array(primitive), z.array(payload), z.array(list)]));
  const value: z.Schema<Value> = z.lazy(() => z.union([primitive, list, payload]));
  const payload: z.Schema<Payload> = z.lazy(() => z.record(value));

  export const schema = payload;
  export const parse = (value: unknown): Payload => schema.parse(value);
}

export namespace EmptyPayload {
  export const schema = z.record(z.never());
}

export type AggregateId<Type extends string = string> = {
  type: Type;
  key: string;
};

export namespace AggregateId {
  export const schema = <Type extends string>(type: Type): z.Schema<AggregateId<Type>> =>
    z.object({
      type: z.literal(type),
      key: z.string().min(1),
    }) as unknown as z.Schema<AggregateId<Type>>;

  export const encode = (aggregate: AggregateId): string => `${aggregate.type}.${aggregate.key}`;
  export const decoder =
    (type: string) =>
    (aggregate: string): AggregateId => {
      const [_type, ...key] = aggregate.split(".");
      return schema(type).parse({ type: _type, key: key.join(".") });
    };

  export const parser =
    <Type extends string>(type: Type) =>
    (value: unknown) =>
      schema(type).parse(value);

  export function create<const T extends string>(type: T, key: string): AggregateId<T> {
    return { type, key } as const;
  }
}

export interface DomainEvent<Type extends string = string, Data extends Payload = Payload> {
  type: Type;
  data: Data;
}

export namespace DomainEvent {
  export const schema = <S extends z.Schema>(type: string, data: S) =>
    z.object({
      type: z.literal(type),
      data,
    });
}

export interface RecordedEvent<E extends DomainEvent = DomainEvent> {
  /**
   * The aggregate that events belongs to.
   */
  aggregate: AggregateId;

  /**
   * The revision for the aggregate.
   */
  revision: string;

  /**
   * Unique identifier representing this event. ULID format.
   */
  id: string;

  /**
   * Type of this event.
   */
  type: E["type"];

  /**
   * Representing when this event was created in the database system.
   */
  timestamp: string;

  /**
   * Data of this event.
   */
  data: E["data"];

  /**
   * Representing the metadata associated with this event.
   */
  metadata: RecordedEvent.Metadata;
}

export namespace RecordedEvent {
  export type Metadata = {
    causationId?: string;
    correlationId?: string;
  };
}

export type Revision = string;

export namespace Revision {
  export type Initial = "00000000000000000000000000";
  export const Initial = "00000000000000000000000000";

  export const isInitial = (revision: Revision): boolean => revision === Initial;

  export const schema = z.string().min(26).max(26);
}
