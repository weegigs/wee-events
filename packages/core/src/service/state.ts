import { Payload } from "../types";

export type State = Payload;
