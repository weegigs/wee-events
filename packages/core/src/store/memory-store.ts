import { DateTime } from "luxon";
import { monotonicFactory } from "ulid";

import {
  AggregateId,
  DomainEvent,
  RecordedEvent,
  Revision,
} from "../types";
import { EventStore } from "./types";

const ulid = monotonicFactory();

export class MemoryStore implements EventStore {
  aggregates: Record<string, RecordedEvent[]> = {};

  clear() {
    this.aggregates = {};
  }

  async load(aggregate: AggregateId): Promise<RecordedEvent[]> {
    return this.aggregates[AggregateId.encode(aggregate)] ?? [];
  }

  async publish(aggregate: AggregateId, events: DomainEvent | DomainEvent[]): Promise<Revision> {
    const current = this.aggregates[AggregateId.encode(aggregate)] ?? [];
    const revision = current[current.length - 1]?.revision ?? Revision.Initial;

    const e = Array.isArray(events) ? events : [events];
    if (e.length == 0) {
      return revision;
    }

    const recorded: RecordedEvent[] = e.map((event) => {
      const revision = ulid()
      const timestamp = DateTime.now().toISO()

      if (timestamp == null) {
        throw new Error("unexpected invalid timestamp")
      }

      return {
        ...event,
        id: revision,
        revision,
        aggregate: aggregate,
        timestamp,
        metadata: {},
      };
    });

    const updated = current.concat(...recorded);
    this.aggregates[AggregateId.encode(aggregate)] = updated;

    return updated[updated.length - 1].revision;
  }
}
