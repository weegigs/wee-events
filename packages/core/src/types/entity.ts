import * as z from "zod";

import { AggregateId, Payload, Revision } from "./core";

export interface Entity<T extends Payload = Payload> {
  aggregate: AggregateId;
  type: string;
  revision: Revision;
  state: T;
}

export namespace Entity {
  export const schema = <Type extends string, Data extends z.ZodTypeAny>(type: Type, state: Data) =>
    z.object({
      aggregate: AggregateId.schema(type),
      type: z.string().min(1),
      revision: Revision.schema,
      state,
    });

  export function create<State extends Payload>(
    aggregate: AggregateId,
    type: string,
    revision: Revision,
    state: State
  ): Entity<State> {
    return { aggregate, type, revision, state } as const;
  }
}
