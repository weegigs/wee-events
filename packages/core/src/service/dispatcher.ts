import * as z from "zod";

import { EventStore } from "../store";
import { Entity, Payload } from "../types";
import { State } from "./state";

export type Command = Payload;

export type Environment = Record<string, unknown>;

export type Publisher = { publish: EventStore.Publisher };

export type Handler<R extends Environment, S extends State, C extends Command> = (
  environment: R,
  entity: Entity<S>,
  command: C
) => Promise<void>;

export type ValidationError = z.ZodError;

export class CommandValidationError extends Error {
  constructor(public readonly error: ValidationError) {
    super("command validation failed");
    this.name = "CommandValidationError";

    Object.setPrototypeOf(this, CommandValidationError.prototype);
  }
}

export class HandlerNotFound extends Error {
  constructor(public readonly path: string) {
    super(`Handler not found for path ${path}`);
    this.name = "HandlerNotFound";

    Object.setPrototypeOf(this, HandlerNotFound.prototype);
  }
}

const validate = <T>(schema: z.Schema<T>, value: unknown): T => {
  const result = schema.safeParse(value);

  if (!result.success) {
    throw new CommandValidationError(result.error);
  }

  return result.data;
};

export interface Dispatcher<S extends State> {
  dispatch(path: string, entity: Entity<S>, command: Command): Promise<void>;
}

export namespace Dispatcher {
  export interface Description {
    commands: Record<string, z.Schema<Payload>>;
  }
}

export interface DispatcherDescription<R extends Environment, S extends State> {
  commands(): Record<string, z.Schema<Payload>>;
  dispatcher(environment: R): Dispatcher<S>;
}

export namespace DispatcherDescription {
  class Factory<R extends Environment, S extends State> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private handlers: Record<string, { schema: z.Schema<any>; handler: Handler<any, S, any> }> = {};

    constructor(
      path: string,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      schema: z.Schema<any>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      handler: Handler<R, S, any>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      existing: Record<string, { schema: z.Schema<any>; handler: Handler<any, S, any> }> = {}
    ) {
      this.handlers = {
        ...existing,
        [path]: { schema, handler },
      };
    }

    handler: <R1 extends Environment, C extends Payload>(
      path: string,
      schema: z.ZodType<C, z.ZodTypeDef, C>,
      handler: Handler<R1, S, C>
    ) => Factory<R & R1, S> = <R1 extends Environment, C extends Payload>(
      path: string,
      schema: z.ZodType<C, z.ZodTypeDef, C>,
      handler: Handler<R1, S, C>
    ) => {
      return new Factory<R & R1, S>(path, schema, handler, this.handlers);
    };

    description: () => DispatcherDescription<R, S> = () => {
      return {
        dispatcher: (environment: R) => ({
          dispatch: async (path: string, entity: Entity<S>, command: Payload) => {
            const handler = this.handlers[path];

            if (!handler) {
              throw new HandlerNotFound(path);
            }

            await handler.handler(environment, entity, validate(handler.schema, command));
          },
        }),
        commands: () => Object.fromEntries(Object.entries(this.handlers).map(([path, { schema }]) => [path, schema])),
      };
    };
  }

  /**
   * Creates a dispatch factory that can be used to create a dispatcher
   *
   * @param path the path to connect the handler to
   * @param schema the zod schema used to validate the command
   * @param handler the handler used to execute the command
   * @returns a dispatch factory that can be used to create a dispatcher
   */
  export function handler<R extends Environment, S extends State, P extends string, C extends Command>(
    path: P,
    schema: z.Schema<C>,
    handler: Handler<R, S, C>
  ): Factory<R, S> {
    return new Factory(path, schema, handler, {});
  }
}

type HF<R extends Environment, P extends string, S extends State, C extends Command> = {
  path: P;
  schema: z.ZodType<C, z.ZodTypeDef, C>;
  handler: Handler<R, S, C>;
};

export function cf<R extends Environment, P extends string, S extends State, C extends Command>(
  hf: HF<R, P, S, C>
): (environment: R, path: P, state: Entity<S>, command: C) => Promise<void> {
  return async (environment: R, _path: P, state: Entity<S>, command: C) => {
    hf.handler(environment, state, validate(hf.schema, command));
  };
}
