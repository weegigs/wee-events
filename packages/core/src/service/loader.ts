import _ from "lodash";
import { z } from "zod";

import * as store from "../store";
import { AggregateId, DomainEvent, Entity, Payload, RecordedEvent, Revision } from "../types";
import { State } from "./state";

export interface EntityLoader<A extends State> {
  load(id: AggregateId): Promise<Entity<A>>;
}

export type EntityDescription<S extends State> = { type: string; schema: z.ZodType<S, z.ZodTypeDef, S> };

export type LoaderDescription<S extends State> = {
  entity(): EntityDescription<S>;
  events(): Record<string, z.Schema>;

  create: (store: store.EventStore) => EntityLoader<S>;
};

export type Reducer<A extends State, Event extends DomainEvent> = (state: A, event: RecordedEvent<Event>) => A;
export type Initializer<A extends State, Event extends DomainEvent> = (event: RecordedEvent<Event>) => A;

export class EntityNotAvailableError extends Error {
  constructor(public readonly aggregate: AggregateId, public readonly type: string) {
    super(`Entity of type ${type} not available from aggregate ${AggregateId.encode(aggregate)}`);
    this.name = "EntityNotAvailableError";

    Object.setPrototypeOf(this, EntityNotAvailableError.prototype);
  }
}

export namespace LoaderDescription {
  type Initializers<A extends State> = Record<
    string,
    { schema: z.Schema<unknown>; initializer: Initializer<A, DomainEvent> }
  >;

  type Reducers<A extends State> = Record<string, { schema: z.Schema<unknown>; reducer: Reducer<A, DomainEvent> }>;

  class Components<A extends State> {
    constructor(
      readonly entity: EntityDescription<A>,
      readonly init: ((aggregate: AggregateId) => A) | undefined,
      readonly initializers: Initializers<A>,
      readonly reducers: Reducers<A>
    ) {}
  }

  const $create = <A extends State>(components: Components<A>): LoaderDescription<A>["create"] => {
    return (store: store.EventStore) => {
      const {
        init,
        initializers,
        reducers,
        entity: { type },
      } = components;

      const replay = (events: RecordedEvent[], state?: A): [A | undefined, Revision] => {
        let result = state;
        let revision = Revision.Initial;

        for (const event of events) {
          if (result === undefined) {
            const { initializer } = initializers[event.type] ?? { initializer: undefined };
            if (initializer !== undefined) {
              result = initializer(event);
              if (result !== undefined) {
                revision = event.revision;
              }
            }
          } else {
            const { reducer } = reducers[event.type] ?? { reducer: undefined };
            if (reducer !== undefined) {
              result = reducer(result, event);
              revision = event.revision;
            }
          }
        }

        return [result, revision];
      };

      const load = async (id: AggregateId) => {
        const events = await store.load(id);
        const [state, revision] = replay(events, typeof init == "function" ? init(id) : undefined);
        if (state === undefined) {
          throw new EntityNotAvailableError(id, type);
        }
        const entity: Entity<A> = {
          aggregate: id,
          type,
          state,
          revision,
        };

        return entity;
      };

      return { load };
    };
  };

  function $events<A extends State>(components: Components<A>): Record<string, z.Schema> {
    const { initializers, reducers } = components;

    return _.merge(
      {},
      ...Object.entries(initializers).map(([event, { schema }]) => ({ [event]: schema })),
      ...Object.entries(reducers).map(([event, { schema }]) => ({ [event]: schema }))
    );
  }

  function $make<A extends State>(components: Components<A>): LoaderDescription<A> {
    const { entity } = components;

    const create = $create(components);
    const events = $events(components);

    return { entity: () => entity, events: () => events, create };
  }

  interface Initialized<A extends State> {
    description(): LoaderDescription<A>;
    reducer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.Schema<Data>,
      reducer: Reducer<A, DomainEvent<Type, Data>>
    ): Initialized<A>;
  }

  class $Initialized<A extends State> extends Components<A> implements Initialized<A> {
    reducer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.Schema<Data>,
      reducer: Reducer<A, DomainEvent<Type, Data>>
    ): Initialized<A> {
      return new $Initialized<A>(this.entity, this.init, this.initializers, {
        ...this.reducers,
        [event]: { schema, reducer: reducer as Reducer<A, DomainEvent> },
      });
    }

    description(): LoaderDescription<A> {
      return $make(this);
    }
  }

  interface Initializing<A extends State> extends Initialized<A> {
    initializer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.ZodType<Data, z.ZodTypeDef, Data>,
      initializer: Initializer<A, DomainEvent<Type, Data>>
    ): Initializing<A>;
  }

  class $Initializing<A extends State> extends $Initialized<A> implements Initializing<A> {
    initializer<Type extends string, Data extends Payload>(
      event: Type,
      schema: z.ZodType<Data, z.ZodTypeDef, Data>,
      initializer: Initializer<A, DomainEvent<Type, Data>>
    ): $Initializing<A> {
      return new $Initializing<A>(
        this.entity,
        this.init,
        {
          ...this.initializers,
          [event]: { schema, initializer: initializer as Initializer<A, DomainEvent> },
        },
        this.reducers
      );
    }
  }

  export const fromInitFunction = <A extends State>(
    entity: { type: string; schema: z.ZodType<A, z.ZodTypeDef, A> },
    init: (aggregate: AggregateId) => A
  ): Initialized<A> => {
    return new $Initialized(entity, init, {}, {});
  };

  export const fromInitializer = <A extends State, Type extends string, Data extends Payload>(
    entity: { type: string; schema: z.ZodType<A, z.ZodTypeDef, A> },
    event: Type,
    schema: z.ZodType<Data, z.ZodTypeDef, Data>,
    initializer: Initializer<A, DomainEvent<Type, Data>>
  ): Initializing<A> => {
    return new $Initializing<A>(
      entity,
      undefined,
      {
        [event]: { schema, initializer: initializer as Initializer<A, DomainEvent> },
      },
      {}
    );
  };
}
