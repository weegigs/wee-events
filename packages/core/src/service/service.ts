import { nanoid } from "nanoid";

import { EventStore } from "../store";
import { AggregateId, DomainEvent, Entity, Payload } from "../types";
import { Command, DespatcherDescription as DispatcherDescription, Environment } from "./dispatcher";
import { EntityDescription, LoaderDescription } from "./loader";
import { State } from "./state";

export type ServiceInfo<S extends Payload> = {
  version: string;
  title?: string;
  description?: string;
  entity: EntityDescription<S>;
};

export type Service<S extends State> = {
  // TODO: KAO - execute needs to be enhanced with a context for cross cutting concerns like authorization.
  execute: (name: string, target: AggregateId, command: Command) => Promise<Entity<S>>;
  load: (aggregate: AggregateId) => Promise<Entity<S>>;
};

export interface ServiceDescription<R extends Environment, S extends State> {
  info(): ServiceInfo<S>;

  commands: DispatcherDescription<R, S>["commands"];
  events: LoaderDescription<S>["events"];
  service(store: EventStore, environment: Omit<R, "publish">): Service<S>;
}

export namespace ServiceDescription {
  class TrackingPublisher {
    triggered = false;

    constructor(
      private store: EventStore,
      private command: string
    ) {}

    publish: EventStore.Publisher = async (
      aggregate: AggregateId,
      events: DomainEvent | DomainEvent[],
      options?: EventStore.PublishOptions
    ) => {
      const correlationId = options?.correlationId ?? `${this.command}/${nanoid(26)}`;
      const result = await this.store.publish(aggregate, events, { ...options, correlationId });
      this.triggered = true;

      return result;
    };
  }

  export function create<R extends Environment, S extends State>(
    info: {
      version: string;
      title?: string;
      description: string;
    },
    loader: LoaderDescription<S>,
    dispatcher: DispatcherDescription<R, S>
  ): ServiceDescription<R, S> {
    return {
      info: () => ({
        ...info,
        entity: loader.entity(),
      }),
      commands: dispatcher.commands,
      events: loader.events,
      service: (store: EventStore, environment: Omit<R, "publish">): Service<S> => {
        const load = loader.create(store).load;
        const execute = async (path: string, target: AggregateId, command: Command) => {
          const tracking = new TrackingPublisher(store, path);
          const dispatch = dispatcher.dispatcher({
            ...environment,
            publish: tracking.publish,
          } as unknown as R).dispatch;

          const entity = await load(target);

          await dispatch(path, entity, command);

          return tracking.triggered ? await load(target) : entity;
        };

        return {
          execute,
          load,
        };
      },
    };
  }
}
