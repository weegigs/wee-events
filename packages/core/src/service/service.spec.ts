import { describe, it, expect, beforeEach } from "vitest";
import { z } from "zod";

import { MemoryStore } from "../store";
import { DomainEvent, Entity } from "../types";
import { DispatcherDescription, Publisher } from "./dispatcher";
import { LoaderDescription, Reducer } from "./loader";
import { ServiceDescription } from "./service";

namespace Calculator {
  export const schema = z.object({
    total: z.number(),
  });
}
type Calculator = z.infer<typeof Calculator.schema>;

namespace Commands {
  export namespace Add {
    export const schema = z.object({
      amount: z.number(),
    });
    export const create = (amount: number): Add => schema.parse({ amount });

    export const handler = async (environment: Publisher, entity: Entity<Calculator>, command: Add) => {
      const { publish } = environment;
      if (command.amount != 0) {
        await publish(entity.aggregate, { type: "added", data: { amount: command.amount } });
      }
    };
  }
  export type Add = z.infer<typeof Add.schema>;

  export namespace Subtract {
    export const schema = z.object({
      amount: z.number(),
    });

    export const create = (amount: number): Subtract => schema.parse({ amount });

    export const handler = async (environment: Publisher, entity: Entity<Calculator>, command: Subtract) => {
      const { publish } = environment;

      if (command.amount != 0) {
        await publish(entity.aggregate, { type: "subtracted", data: { amount: command.amount } });
      }
    };
  }

  export type Subtract = z.infer<typeof Subtract.schema>;

  export namespace Clear {
    export const schema = z.object({});

    export const create = (): Clear => schema.parse({});

    export const handler = async (environment: Publisher, entity: Entity<Calculator>, _: Clear) => {
      const { publish } = environment;

      if (entity.state.total != 0) {
        await publish(entity.aggregate, { type: "cleared", data: {} });
      }
    };
  }
  type Clear = z.infer<typeof Clear.schema>;
}

namespace Events {
  export namespace Added {
    export const schema = DomainEvent.schema("added", z.object({ amount: z.number() }));

    export const create = (amount: number): Added => schema.parse({ type: "added", data: { amount } });

    export const reducer: Reducer<Calculator, Added> = (state, event) => {
      return { ...state, total: state.total + event.data.amount };
    };
  }
  export type Added = z.infer<typeof Added.schema>;

  export namespace Subtracted {
    export const schema = DomainEvent.schema("subtracted", z.object({ amount: z.number() }));
    export const create = (amount: number): Subtracted => schema.parse({ type: "subtracted", data: { amount } });
    export const reducer: Reducer<Calculator, Subtracted> = (state, event) => {
      return { ...state, total: state.total - event.data.amount };
    };
  }
  export type Subtracted = z.infer<typeof Subtracted.schema>;

  export namespace Cleared {
    export const schema = DomainEvent.schema("cleared", z.object({ amount: z.number() }));
    export const create = (amount: number): Cleared => schema.parse({ type: "cleared", data: { amount } });
    export const reducer: Reducer<Calculator, Cleared> = (state) => {
      return { ...state, total: 0 };
    };
  }
  export type Cleared = z.infer<typeof Cleared.schema>;
}

describe("entity service", () => {
  const store = new MemoryStore();
  const aggregate = { type: "test", key: "1" };

  const loader = LoaderDescription.fromInitFunction<Calculator>(
    { type: "calculator", schema: Calculator.schema },
    () => ({
      total: 0,
    })
  )
    .reducer("added", Events.Added.schema.shape.data, Events.Added.reducer)
    .reducer("subtracted", Events.Subtracted.schema.shape.data, Events.Subtracted.reducer)
    .reducer("cleared", Events.Cleared.schema.shape.data, Events.Cleared.reducer)
    .description();

  const dispatcher = DispatcherDescription.handler("add", Commands.Add.schema, Commands.Add.handler)
    .handler("subtract", Commands.Subtract.schema, Commands.Subtract.handler)
    .handler("clear", Commands.Clear.schema, Commands.Clear.handler)
    .description();

  const description = ServiceDescription.create(
    { title: "test", description: "test service", version: "1.0.0" },
    loader,
    dispatcher
  );

  const service = description.service(store, {});

  beforeEach(() => {
    store.clear();
  });

  it("should load an empty entity from the store", async () => {
    const entity = await service.load(aggregate);

    expect(entity.state).toEqual({ total: 0 });
    expect(entity).toMatchSnapshot();
  });

  it("should load an entity with existing state", async () => {
    store.publish(aggregate, Events.Added.create(10));
    store.publish(aggregate, Events.Subtracted.create(3));
    const entity = await service.load(aggregate);

    expect(entity.state).toEqual({ total: 7 });
  });

  it("should return an updated value when a command is executed", async () => {
    const entity = await service.execute("add", aggregate, Commands.Add.create(10));
    expect(entity.state).toEqual({ total: 10 });
  });

  it("should maintain state between command executions", async () => {
    let entity = await service.execute("add", aggregate, Commands.Add.create(10));
    expect(entity.state).toEqual({ total: 10 });

    entity = await service.execute("subtract", aggregate, Commands.Subtract.create(7));
    expect(entity.state).toEqual({ total: 3 });

    entity = await service.execute("clear", aggregate, Commands.Clear.create());
    expect(entity.state).toEqual({ total: 0 });

    const result = await store.load(aggregate);

    expect(result).toHaveLength(3);
  });
});