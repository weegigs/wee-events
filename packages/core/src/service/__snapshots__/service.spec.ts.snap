// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity service should load an empty entity from the store 1`] = `
{
  "aggregate": {
    "key": "1",
    "type": "test",
  },
  "revision": "00000000000000000000000000",
  "state": {
    "total": 0,
  },
  "type": "calculator",
}
`;
